### PR TITLE
Add metrics sidebar

### PR DESCRIPTION
**Pull request by hellodommy**
_(see comment below)_

### CODE REVIEW

**Comment by kenneth-fung** [comment]
Metrics ready on Settings page. Will integrate into Dashboard once merged.

image

- Degree Progress updates based on the currently selected plan.
- CAP doesn't update, but functions to calculate it are in place (not used since the bit of code that adds actual/projected grade to mods is in the `dashboard` branch)
- Degree Requirements updates based on the current plan. All buckets filled with CS Degree mods. Mods are color-coded in the following manner (based on Degree Progress legend):
  - Primary Purple: Completed
  - Light Purple: Planned
  - Red: Unallocated and needed to fulfill degree requirements
  - Grey: Unplanned but not necessary to fulfill degree requirements

File: src/components/Metrics/DegreeRequirements/fillDegreeReqData.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            } else if (Object.keys(buckets.itProf).includes(mod.code)) {
+                // IT Professionalism
+                buckets.itProf[mod.code] = getAllocatedModState(sem);
+            } else if (Object.keys(buckets.mathSci).includes(mod.code)) {
+                // Math & Science
+                buckets.mathSci[mod.code] = getAllocatedModState(sem);
+            } else {
+                // check if mod is a science mod
+                const department = mod.code.substring(0, 2);
+                if (department === "CM" || department === "PC" || department + "M" === "LSM") {
```
Comment (hellodommy) [inline src/components/Metrics/DegreeRequirements/fillDegreeReqData.js:33]: wow

File: src/utils/calculateCaps.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        case "C":
+            return 2.0;
+        case "D+":
+            return 1.5;
+        case "D+":
+            return 1.0;
+        case "F":
+            return 0.0;
+        default:
+            return "THIS AIN'T A GRADE!";
```
Comment (hellodommy) [inline src/utils/calculateCaps.js:26]: LOL

File: src/components/Metrics/Metrics.jsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  const { bndMcs, fourKMcs } = fillDegreeReqData(plan);
+  const { numCompleted, numPlanned, numUnallocated } = calculateProgress(plan.specialisations);
+  const total = numCompleted + numPlanned + numUnallocated;
+
+  const capTooltip = (
+    <Popover className="metrics-tooltip">
+      <Popover.Content>
+        A student's CAP is the sum of the module grade points multiplied by the number of MCs for
+        the corresponding module, divided by the total number of MCs.<br/><br/>
+        Visit <a href="http://www.nus.edu.sg/registrar/academic-information-policies/non-graduating/modular-system">this page</a> for
```
Comment (flxffy) [inline src/components/Metrics/Metrics.jsx]:
```suggestion
        Visit <a href="http://www.nus.edu.sg/registrar/academic-information-policies/non-graduating/modular-system" target="_blank" rel="noopener noreferrer">this page</a> for
```

File: src/utils/calculateCaps.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        case "B-":
+            return 3.0;
+        case "C+":
+            return 2.5;
+        case "C":
+            return 2.0;
+        case "D+":
+            return 1.5;
+        case "D+":
+            return 1.0;
```
Comment (flxffy) [inline src/utils/calculateCaps.js]:
```suggestion
        case "D":
            return 1.0;

```

**Review by flxffy — CHANGES_REQUESTED** [review]
Typo! Otherwise everything looks good :D

File: src/utils/calculateCaps.js (hunk 4 above)
Comment (kenneth-fung) [reply src/utils/calculateCaps.js]: Oh god, thanks for catching.

**Pull request merged**